courier/courier_dispatcher.go

```diff
@@ -77,7 +61,27 @@ func (c *courier) DispatchQueue(ctx context.Context) error {
 	}
 
 	for k, msg := range messages {
-		if err := c.DispatchMessage(ctx, msg); err != nil {
+		if msg.SendCount > maxRetries {
+			if err := c.deps.CourierPersister().SetMessageStatus(ctx, msg.ID, MessageStatusAbandoned); err != nil {
+				c.deps.Logger().
+					WithError(err).
+					WithField("message_id", msg.ID).
+					Error(`Unable to set the retried message's status to "abandoned".`)
+				return err
+			}
+			// Skip the message
+			c.deps.Logger().
+				WithField("message_id", msg.ID).
+				Warnf(`Message was abandoned because it did not deliver after %d attempts`, msg.SendCount)
+
+		} else if err := c.DispatchMessage(ctx, msg); err != nil {
+
+			if err := c.deps.CourierPersister().RecordDispatch(ctx, msg.ID, CourierMessageDispatchStatusFailed, err.Error()); err != nil {
```

Comment:
`err.Error()` will be pretty meaningless, it will show something like "Internal Server Error". I think it makes more sense to serialize this to JSON (as we do when sending error responses in APIs) and record that.

Reply:
`err.Error()` returns `dial tcp: lookup my.custom.mailserver.host on 127.0.0.11:53: no such host` for example. I think that's all the errors users need, right?

> I think it makes more sense to serialize this to JSON 

Hm, interesting. 

How do we display them then? Just showing plain JSON is not great...

I updated this PR and will try it out in the network next. Good idea!

Reply:
You can still parse the JSON in the UI :)